mavis-aggregator/src/index.ts

```diff
@@ -0,0 +1,136 @@
+import {Hono} from 'hono'
+import Redis from "ioredis";
+
+const MAX_BODY_SIZE = 1e6; // 1MB in bytes
+const MAX_REQUESTS_PER_SECOND = 100;
+const FLUSH_INTERVAL = 1000; // 1 second in milliseconds
+const MAVIS_API_URL = "https://x.skymavis.com/track";
+
+type Event = {
+    type: string,
+    data: Record<string, any>,
+}
+
+type QueuedEvent = {
+    api_key: string;
+    event: Event,
+}
+
+type TrackEventBody = {
+    api_key: string;
+    events: Event[],
+}
+
+let requestsInLastSecond = 0;
+let lastRequestTime = Date.now();
+const redis = new Redis(Number(process.env.REDIS_PORT ?? 6379), process.env.REDIS_HOST!);
+
+const app = new Hono();
+const calculatePayloadSize = (events: Event[]): number => {
+    return new TextEncoder().encode(JSON.stringify(events)).length;
+}
+
+const flushEvents = async (): Promise<void> => {
+    const now = Date.now();
+    if (now - lastRequestTime >= FLUSH_INTERVAL) {
+        requestsInLastSecond = 0;
+        lastRequestTime = now;
+    }
+
+    if (requestsInLastSecond >= MAX_REQUESTS_PER_SECOND) {
+        console.log('Rate limit reached, deferring flush');
+        setTimeout(flushEvents, FLUSH_INTERVAL);
+        return;
+    }
+
+    let eventsToSend: QueuedEvent[] = [];
+    let currentSize = 0;
+
+    while (true) {
```

Comment:
I'm more or less sure the size limit is per request -- so this needs to happen AFTER grouping the events. 

Similar to the rate limit, it likely exist per API key

Reply:
The fact that we are using TS doesn't mean we need to throw away what our lord .NET taught us. 

Keep this simple. Make a class for handling the event sending per api key